### PR TITLE
Check if second run of rule is not fired

### PR DESCRIPTION
**Pull request by kokosing**
Check if second run of rule is not fired

It seems natural that optimizer Rule if run on output of the same Rule
should not fire again.
While we may think of rules that do only part of work on first run, and
some continuation of work on the second run, it does not seem like
necessary feature.

### CODE REVIEW

**Comment by kokosing** [comment]
See also https://github.com/Teradata/presto/issues/499

**Comment by findepi** [comment]
> It seems natural that optimizer Rule if run on output of the same Rule should not fire again.

I think such _stepping_ was proposed already. E.g. when we were discussing empty values optimization #7292

**Comment by martint** [comment]
I had a check like that in one of my initial prototypes. It doesn't work because it breaks cases like:

- limit
  - limit 
     - limit

when applying a rule like MergeLimits.

**Comment by kokosing** [comment]
I see your point. In general we cannot have such check, but still in tests in might be valuable to have it. It is easy to write a recursive rule. We already had such issue when rule was recursive in some circumstances and it was not found until the code review, tests passed successfully.

As recursive rule I mean, when rule for example from:
```
 - Filter
   - AssignUniqueId
```

creates:
``` 
 - Filter
   - AssignUniqueId
     - Filter
```

and then it can still do (in a loop):
``` 
 - Filter
   - AssignUniqueId
     - Filter
       - Filter
```


I will refactor this in a way that user (developer) can say if he expect that this rule should apply once or more times in a given test.

**Comment by findepi** [comment]
I agree we should not expect rules to be inapplicable to their output. But, we can surely expect rules not to be _infinitely recursive_. In practice, this can be tested by limiting number of times a rule is applied (in tests). The `RuleTester` can try-apply the rule e.g. max 10 times and fail if it applied every time.

**Comment by kokosing** [comment]
I changed the approach. Now developer can in tests how many times it is expected given rule can be applied to given plan.

**Comment by kokosing** [comment]
Rebased

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java [hunk 1]
```diff
@@ -73,10 +76,18 @@ public void doesNotFire()
...
+
+        return this;
     }
 
-    public void matches(PlanMatchPattern pattern)
+    /**
+     * Assert that rule is applied, if so then plan stored in this {@link RuleAssert} will be updated
+     * to the one returned by rule.
+     */
+    public RuleAssert isFiredAndThen()
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java]: Why do we need this method? `matches()` should imply that the rule fires.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java [hunk 2]
```diff
@@ -106,6 +117,35 @@ public void matches(PlanMatchPattern pattern)
...
         }
 
-        assertPlan(session, metadata, new Plan(actual, symbolAllocator), pattern);
+        plan = actual;
+        symbols = symbolAllocator.getTypes();
+
+        return this;
+    }
+
+    public RuleAssert isFiredOnlyOnceAndThen()
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java]: I'd call this `firesOnce()`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java [hunk 3]
```diff
@@ -106,6 +117,35 @@ public void matches(PlanMatchPattern pattern)
...
+    {
+        return isFiredMultipleTimesAndThen(1);
+    }
+
+    public RuleAssert isFiredMultipleTimesAndThen(int count)
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java]: I'm not 100% convinced we want to extend the framework to allow these kinds of iterative tests. We miss the ability to test the intermediate transformations, for instance. My original intent was for the rule assertion framework to verify single rule applications.

I do think it's important to test iterative application of rules, but that can be done with the logical planner tests (TestLogicalPlanner and friends), or by building something like it for sets of rules.

**Comment by kokosing** [comment]
Closing this as it does not seem to be merged.

**Pull request closed without merging**